### PR TITLE
Insert dob and gender too fix #2286

### LINKED ISSUE #2286: import_persons.php basically only uses names when importing data
See the where condition here: https://github.com/thewca/worldcubeassociation.org/blob/abc382dbcd338bd83c599b1640a8e005ac042305/webroot/results/admin/scripts/import_persons.php#L33-L34

It's only matching on `name`, `countryId`, and is restricting the search to people who competed in this particular competition.

We run `import_persons.php` *after* running `persons_finish_unfinished.php`, so it should be possible to match on WCA ID here. In order for this to be possible, we'll need to update `persons_finish_unfinished` to update the `InboxPersons` as well as the `Persons` and `Results` tables here:

https://github.com/thewca/worldcubeassociation.org/blob/abc382dbcd338bd83c599b1640a8e005ac042305/webroot/results/admin/persons_finish_unfinished_ACTION.php#L58-L62

### PR DESCRIPTION
**Pull request by Baiqiang**
@jfly I have a better idea here. We fetch the dob and gender from the `InboxPersons` table and then insert into the `Persons` table. So we don't need to update any missing data since there should be none.
I've tested it on staging, but @SAuroux can you also test it?

### CODE REVIEW

**Comment by SAuroux** [comment]
Hm, as far as I understand you have now simply merged the data import step into the newcomer creation, but you did not address the problem at all? The code now only goes name and countryId, not even by competitionId anymore which creates a new potential (yet very unlikely) bug when posting multiple competitions in parallel. Or am I missing something?

Also, what happens if you run persons_finish_unfinished for a person with no data in InboxPersons? This happens when we split profiles. I just tried this and got this:

image

**Comment by Baiqiang** [comment]
@SAuroux Thanks for the feedback.
I know your question when I was writing the codes. In the `import_persons` script, it has param `compId`. But did you realize that when we are creating new comers, we also update the `personId` in `Results` table according to name and countryId only? This means if you're posting multiple results in parallel and there're more than one new comer with same name and countryId, the first one will be given a new WCA ID and the rest ones in other competitions will be updated to the same WCA ID. See here: https://github.com/thewca/worldcubeassociation.org/blob/abc382dbcd338bd83c599b1640a8e005ac042305/webroot/results/admin/persons_finish_unfinished_ACTION.php#L183-L187. So I assumed  we will finish creating new comer one competition after one instead of in parallel.

For the question that no data in InboxPersons, that's my fault and I'll fix it soon.

**Comment by Baiqiang** [comment]
@SAuroux I've updated it. Please test it on staging (maybe you should switch to the branch fix-2286 if it's not on this branch), thanks!

**Comment by SAuroux** [comment]
Hey Baiqiang,

I have just finished testing multiple scenarios:

- I uploaded the results of Ha Noi Open 2017 that caused problems back then. Everything worked perfectly!
- I also uploaded another competition's results in parallel, to ensure that nothing breaks there. Since the combination of personId and competitionId remains unique, everything is still perfect!

However, there is still a bug with empty personIds:

image

I believe that I have found the wrong line of code causing this (see review comment).

But moreover, this is quite a groundbreaking change! I think I should not be the only person to look over this before this gets deployed on production. @jfly?

File: webroot/results/admin/persons_finish_unfinished_ACTION.php [hunk 1]
```diff
@@ -31,67 +31,89 @@ function finishUnfinishedPersons () {
...
     $action = getRawParamThisShouldBeAnException( "action$caseNr" );
 
+    $person = null;
+    if ( $personId ) {
+      $person = getInboxPerson( $personId, $competitionId );
+    }
+    if ( $person === null ) {
+      $person = array(
+        'name' => $oldName,
+        'countryId' => $countryId,
```
Comment (SAuroux) [inline webroot/results/admin/persons_finish_unfinished_ACTION.php]: I think this needs to be `'countryId' => $oldCountry,`

**Comment by Baiqiang** [comment]
@SAuroux Thanks for testing. You're right about the bug. I've fixed it. Let's wait for @jfly on this.

File: webroot/results/admin/persons_check_finished.php [hunk 2]
```diff
@@ -183,20 +183,20 @@ function checkTooMuchInResults () {
 #----------------------------------------------------------------------
   global $personsFromPersons, $personsFromResults;
   echo "<hr />";
-  
+
   $tooMuchInResults = array();
   #--- Find all ('finished') entries in Results that don't have a match in Persons.
   foreach( array_keys( $personsFromResults ) as $personKey ){
-    if( $personsFromResults[$personKey]['id']  &&  ! isset($personsFromPersons[$personKey]) )
+    if( $personsFromResults[$personKey]['id'] && !ctype_digit($personsFromResults[$personKey]['id']) &&  ! isset($personsFromPersons[$personKey]) )
```
Comment (jfly) [inline webroot/results/admin/persons_check_finished.php]: It took me a while to figure out what's going on here. Could you add a comment here explaining why person ids can be all numeric now?

File: webroot/results/admin/persons_finish_unfinished.php [hunk 3]
```diff
@@ -60,14 +60,18 @@ function getPersons () {
...
-    JOIN Competitions ON Competitions.id=Results.competitionId
-    WHERE personId=''
+    LEFT JOIN Competitions ON Competitions.id=Results.competitionId
+    LEFT JOIN InboxPersons ON InboxPersons.id=Results.personId and InboxPersons.competitionId=Results.competitionId
+    WHERE personId='' OR personId REGEXP '^[0-9]+$'
+    GROUP BY Results.personId, Results.personName, Results.competitionId
     ORDER BY personName
   ");
-  while( $row = mysql_fetch_row( $persons ))
+  while( $row = mysql_fetch_array( $persons ))
```
Comment (jfly) [inline webroot/results/admin/persons_finish_unfinished.php]: Just curious: why the change here? I'm no PHP guy, but https://stackoverflow.com/a/11480149 says "It is recommended to use either _assoc or _row"?

File: webroot/results/admin/scripts/import_results.php [hunk 4]
```diff
@@ -5,9 +5,10 @@
 $compId = getNormalParam('c');
 
 // move over results
+// insert personId for new comers
```
Comment (jfly) [inline webroot/results/admin/scripts/import_results.php]: nit: newcomers is one word

File: webroot/results/admin/persons_finish_unfinished.php [hunk 5]
```diff
@@ -178,11 +184,14 @@ function showUnfinishedPersons () {
...
     $countryIdHtml = htmlEscape( $countryId );
+    $personId = htmlEscape( $person['personId'] );
+    $competitionId = htmlEscape( $person['competitionId'] );
+    $dob = empty($person['dob']) ? 'mm/dd/yyyy' : $person['dob'];
 
     #--- Hidden field describing the case.
     $caseNr++;
-    tableRowFull( "&nbsp;<input type='hidden' name='oldNameAndCountry$caseNr' value='$nameHtml|$countryIdHtml' />" );
-    
+    tableRowFull( "&nbsp;<input type='hidden' name='oldNameAndCountry$caseNr' value='$nameHtml|$countryIdHtml|$personId|$competitionId' />" );
```
Comment (jfly) [inline webroot/results/admin/persons_finish_unfinished.php]: This parameter name seems inaccurate now? Should this be `oldNameAndCountryAndPersonIdAndCompId`? ;)

File: webroot/results/admin/persons_finish_unfinished_ACTION.php [hunk 6]
```diff
@@ -31,67 +31,89 @@ function finishUnfinishedPersons () {
...
 
+    $person = null;
+    if ( $personId ) {
+      $person = getInboxPerson( $personId, $competitionId );
+    }
+    if ( $person === null ) {
+      $person = array(
+        'name' => $oldName,
+        'countryId' => $oldCountry,
+        'competitionId'=>$competitionId,
```
Comment (jfly) [inline webroot/results/admin/persons_finish_unfinished_ACTION.php]: nit: missing spaces around the `=>`

File: webroot/results/admin/persons_finish_unfinished.php (hunk 3 above)
Comment (Baiqiang) [reply webroot/results/admin/persons_finish_unfinished.php]: I don't know why they recommand use _assoc or _row instead of _array. In PHP doc it says:
>An important thing to note is that using mysql_fetch_array() is not significantly slower than using mysql_fetch_row(), while it provides a significant added value.
